File: csharp/880.decoded-string-at-index.cs
/*
 * @lc app=leetcode id=880 lang=csharp
 *
 * [880] Decoded String at Index
 */

// @lc code=start
public class Solution {
    public string DecodeAtIndex(string S, int K) {
        long count = 0, l = (long)K;
        int i;
        for (i = 0; count < l; ++i) {
            char c = S[i];
            if ('a' <= c && c <= 'z') {
                ++count;
            } else {
                count *= (long)((int)c - (int)'0');
            }
        }
        for (--i; i >= 0; --i) {
            char c = S[i];
            if ('a' <= c && c <= 'z') {
                if (l % count == 0) {
                    return new string(c, 1);
                }
                --count;
            } else {
                count /= (long)((int)c - (int)'0');
                l %= count;
            }
        }
        return "";
    }
}
// @lc code=end

File: java/880.decoded-string-at-index.java
/*
 * @lc app=leetcode id=880 lang=java
 *
 * [880] Decoded String at Index
 */

// @lc code=start
class Solution {
    public String decodeAtIndex(String S, int K) {
        long len = 0, r = (long)K;
        int i, n = S.length();
        for (i = 0; len < r; ++i) {
            char c = S.charAt(i);
            if (c <= 'z' && c >= 'a') {
                len++;
            } else {
                len *= (long)c - (long)'0';
            }
        }
        for (--i; i >= 0; --i) {
            char c = S.charAt(i);
            if (c <= 'z' && c >= 'a') {
                if (r % len == 0) {
                    return Character.toString(c);
                }
                --len;
            } else {
                len /= (long)c - (long)'0';
                r %= len;
            }
        }
        return "";
    }
}
// @lc code=end

